Add bundling with rollup to docs

DonNicoJs opened the pull request:
This PR adds a new block to the `single-file-component.md` file with a guide on how to bundle `sfc` with rollup.

The guide is oriented to library authors.

This is the first draft, also I am not completely sure it's the right place to add this, or if is better somewhere else / standalone entry

fixes #660

DonNicoJs commented:
@NataliaTepluhina This is a first draft, I believe is quite rough and it needs some iteration to be presentable 😅 

I was thinking that a boilerplate repository is probably a good addition ( as suggested by @JessicaSachs ) where such a repo should 'live' ? Shall I create it under my personal repositories or there is a better place for such things?

In the end I went quite minimal with the approach and discarded the composition api point completely since I think it is a bit out of the scope.

DonNicoJs commented:
@znck Currently on my side I still have some quirks with the render functions, I was thinking to add this to a possible tips and trick section, but then I realised that RPV is still in beta ( I am still using the alpha ) so it probably does not make much sense to add things that are planned to be fixed ( or already fixed and not released )

znck commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,108 @@ After you've taken a day to dive into these resources, we recommend checking out
+We will need to install rollup and a few dependencies:
+
+```bash
+npm install --save-dev rollup rollup-plugin-commonjs rollup-plugin-vue 
````
Commonjs plugin is moved to `@rollup/plugin-commonjs`

DonNicoJs replied:
Updated!

znck commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,108 @@ After you've taken a day to dive into these resources, we recommend checking out
+      name: "MyLibraryName",
+      sourcemap: true,
+      globals: {
+        vue: "vue",
```
I think it should be `vue: "Vue"`.

DonNicoJs replied:
Updated! 🙇

znck commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,108 @@ After you've taken a day to dive into these resources, we recommend checking out
+  output: [ 
+    {
+      file: pkg.module, // the name of our esm librry
+      format: "es", // the format of choice
```
Let's use `"esm"` because we said so in above section. (This is first mention of `"es"`)

DonNicoJs replied:
Done!

znck commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,108 @@ After you've taken a day to dive into these resources, we recommend checking out
+npm install --save-dev rollup rollup-plugin-commonjs rollup-plugin-vue 
+```
+
+The command above will install [rollup](https://webpack.js.org), [rollup-plugin-commonjs](https://github.com/rollup/plugins/tree/master/packages/commonjs) and 
````
We can skip this line and directly start with, "These are the minimal ...."

DonNicoJs replied:
I agree! Updated!

znck reviewed (commented):
This is a good start 🤩

Maybe we can add notes about typescript and vetur support too.

DonNicoJs commented:
> This is a good start 🤩
> 
> Maybe we can add notes about typescript and vetur support too.

@znck  Thanks! I've applied all the suggestions 😄  about Typescript, it falls out of my experience (I know I have to catch up with Typescript 😅 ) And I've never configured Vetur for vue3.

Do you think this is super important to have at the first version of this part of the docs or we can add those in a further iteration? 

@NataliaTepluhina Is there anything else that need to change in. your opinion? WDYT? 🙇 🙇

znck commented:
Typescript stuff can be done in a separate PR. 👍

skirtles-code commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
 
 In case you prefer setting up your own build setup from scratch, you will need to manually configure webpack with [vue-loader](https://vue-loader.vuejs.org). To learn more about webpack itself, check out [their official docs](https://webpack.js.org/configuration/) and [webpack learning academy](https://webpack.academy/p/the-core-concepts).
+
+### Building with rollup
```
I'll put this comment here but it applies more generally to this section.

From a quick look at the Rollup documentation it would appear that they use a capital R on Rollup when referring to their library. The usual convention in the Vue documentation is to adhere to the casing conventions used by a project's own documentation. So, for example, both npm and webpack are typically written in lowercase.

The `rollup` package and `rollup` command are both lowercase, so when referring to them we obviously wouldn't use a capital R. However, both of those uses would usually be surrounded with backticks when included in the middle of a sentence.

DonNicoJs replied:
@skirtles-code thanks! I think I've addressed this feedback along the whole piece!

skirtles-code commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+
+### Building with rollup
+
+Most of the time when developing a third party library we want to build it in a way that allows the consumers of the library to [tree shake it](https://webpack.js.org/guides/tree-shaking/) to achieve so we need to build along with `commonjs` and `umd/iife` modules also `esm` modules. Since webpack and in turn vue-cli do not support building `esm` modules we need to rely on [rollup](https://rollup ...
```
I think 'third-party' should be hyphenated here.

I'd move the word 'it' outside that first link text.

I got lost where it says 'to achieve'. That bit doesn't seem to follow on from the earlier part of the sentence.

DonNicoJs replied:
@skirtles-code Thanks!

skirtles-code commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+npm install --save-dev rollup @rollup/plugin-commonjs rollup-plugin-vue 
+```
+
+These are the minimal amount of rollup plugins that we need to use to compile the code in an `esm` module, users may want to also add [rollup-plugin-babel](https://github.com/rollup/plugins/tree/master/packages/babel) to transpile their code and [node-resolve](https://github.com/rollup/plugins/tree/master/packages/node-resolve) if they use dependencies that they want to bundle with the library.
````
I think the comma after 'module' should be promoted to a full-stop/period.

DonNicoJs replied:
@skirtles-code Thanks! Done!

skirtles-code commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+touch rollup.config.js
+```
+
+Once the file is created we will need to open it with our editor of choice and add the following code
````
There should be a colon at the end of this line.

DonNicoJs replied:
@skirtles-code Thanks! Done!

skirtles-code commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+
+```javascript
+// import our third party plugins
+import commonjs from "rollup-plugin-commonjs";
````
I think I'm right in saying that the Vue documentation uses the following conventions for JS code examples:

1. Single quotes for strings.
2. No semi-colons at the ends of lines.
3. No trailing commas in objects and arrays.

I'm sure there are exceptions but I think most of the examples do follow those conventions.

DonNicoJs replied:
@skirtles-code Thanks! Done!

skirtles-code commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+
+export default {
+  // this is the file containing all our exported components/functions
+  input: 'src/index.js' 
```
I think there's a missing comma here.

DonNicoJs replied:
@skirtles-code Thanks! Done!

skirtles-code commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+
+- how to build our package
+- what files we want to bundle in our package
+- what file represents our esm module
```
Backticks on `esm`?

DonNicoJs replied:
@skirtles-code Thanks! Done!

skirtles-code commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+
+#### Bundling `umd` and `cjs` modules
+
+To build along with `esm` modules also `umd` and `cjs` ones we can simply add a few lines of configuration to our `rollup.config.js` and `package.json`
```
I don't think the start of this sentence quite works.

How about...

```text
To also build `umd` and `cjs` modules we can simply...
```

DonNicoJs replied:
@skirtles-code Thanks! Done!

skirtles-code commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+To build along with `esm` modules also `umd` and `cjs` ones we can simply add a few lines of configuration to our `rollup.config.js` and `package.json`
+
+##### rollup.config.js 
+```javasript
````
Typo.

DonNicoJs replied:
@skirtles-code Thanks! Done! 🤦

skirtles-code reviewed (changes requested):
I don't really know anything about Rollup but I've had a read through and made a few notes on anything that caught my eye. All minor stuff.

I see some similarities with #664, where @sdras is migrating some Rollup-related content in the Cookbook. I don't know enough about the topic myself to give an opinion but I wonder whether the two need to be considered together?

DonNicoJs commented:
@@skirtles-code Thank you so much for the review! And apologies for the slow turn-around, I manage to scavenge time for OSS only during the weekend 🙇  I think I've addressed all the points!

skirtles-code commented on src/guide/single-file-component.md:
```diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+
+### Building with rollup
+
+Most of the time when developing a third-party library we want to build it in a way that allows the consumers of the library to [tree shake](https://webpack.js.org/guides/tree-shaking/) it. To enable three-shaking we need to build along with `commonjs` and `umd/iife` modules also `esm` modules. Since webpack and in turn vue-cli do not support building `esm` modules we need to rely on [rollup](htt ...
```
```suggestion
Most of the time when developing a third-party library we want to build it in a way that allows the consumers of the library to [tree shake](https://webpack.js.org/guides/tree-shaking/) it. To enable tree-shaking we need to build `esm` modules. Since webpack and, in turn, vue-cli do not support building `esm` modules we need to rely on [rollup](https://rollupjs.org/).
```

I'm not sure `commonjs` and `umd` need mentioning here. Including them also makes the grammar more difficult to get right. My suggestion is to remove them but we can rework this if you think they're important.

skirtles-code commented on src/guide/single-file-component.md:
````diff
@@ -65,3 +65,107 @@ After you've taken a day to dive into these resources, we recommend checking out
+npm install --save-dev rollup @rollup/plugin-commonjs rollup-plugin-vue 
+```
+
+These are the minimal amount of rollup plugins that we need to use to compile the code in an `esm` module. Users may want to also add [rollup-plugin-babel](https://github.com/rollup/plugins/tree/master/packages/babel) to transpile their code and [node-resolve](https://github.com/rollup/plugins/tree/master/packages/node-resolve) if they use dependencies that they want to bundle with the library.
````
When you say 'Users' here, who does that refer to? Perhaps I've misunderstood but it seems like that's referring to the reader? The paragraph starts off with 'we' but then switches to the third person. I think we'd want to carry on using we/our throughout.

DonNicoJs replied:
Yeah you are right, it's pointing to the reader, I will go through it and 'move' everything to `we/our`

DonNicoJs replied:
@skirtles-code Thanks! 🙇 This addressed too! And suggestion applied! Please let me know if there is more than I can adjust and again thank you so much for the super thorough review!

skirtles-code reviewed (commented):
A couple more thoughts from me.

I don't have the required permissions to mark my previous comments as resolved but if someone else wants to do that then feel free.

NataliaTepluhina commented:
@DonNicoJs @skirtles-code thank you both so much for authoring the addition and for the thorough review! I believe right now there are no major blockers for merging this one and we can handle additions like TS in the follow-ups

DonNicoJs commented:
@NataliaTepluhina Thank you very much! 🙇 🙇
